src/main/java/org/opensearch/agent/tools/RAGTool.java

```diff
@@ -0,0 +1,275 @@
+import org.opensearch.ml.common.transport.prediction.MLPredictionTaskRequest;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * This tool supports retrieving helpful information to optimize the output of the large language model to answer questions..
+ */
+@Log4j2
+@Setter
+@Getter
+@ToolAnnotation(RAGTool.TYPE)
+public class RAGTool extends AbstractRetrieverTool {
+    public static final String TYPE = "RAGTool";
+    public static String DEFAULT_DESCRIPTION =
+        "Use this tool to retrieve helpful information to optimize the output of the large language model to answer questions.";
+    public static final String INFERENCE_MODEL_ID_FIELD = "inference_model_id";
+    public static final String EMBEDDING_MODEL_ID_FIELD = "embedding_model_id";
+    public static final String EMBEDDING_FIELD = "embedding_field";
+    public static final String OUTPUT_FIELD = "output_field";
+    private String name = TYPE;
+    private String description = DEFAULT_DESCRIPTION;
+    private Client client;
+    private String inferenceModelId;
+    private NamedXContentRegistry xContentRegistry;
+    private String index;
+    private String embeddingField;
+    private String[] sourceFields;
+    private String embeddingModelId;
+    private Integer docSize;
+    private Integer k;
+    @Setter
+    private Parser inputParser;
+    @Setter
+    private Parser outputParser;
+
+    @Builder
+    public RAGTool(
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        String index,
+        String embeddingField,
+        String[] sourceFields,
+        Integer k,
+        Integer docSize,
+        String embeddingModelId,
+        String inferenceModelId
+    ) {
+        super(client, xContentRegistry, index, sourceFields, docSize);
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.index = index;
+        this.embeddingField = embeddingField;
+        this.sourceFields = sourceFields;
+        this.embeddingModelId = embeddingModelId;
+        this.docSize = docSize == null ? DEFAULT_DOC_SIZE : docSize;
+        this.k = k == null ? DEFAULT_K : k;
+        this.inferenceModelId = inferenceModelId;
+
+        outputParser = new Parser() {
+            @Override
+            public Object parse(Object o) {
+                List<ModelTensors> mlModelOutputs = (List<ModelTensors>) o;
+                return mlModelOutputs.get(0).getMlModelTensors().get(0).getDataAsMap().get("response");
+            }
+        };
+    }
+
+    // getQueryBody is not used in RAGTool
+    @Override
+    protected String getQueryBody(String queryText) {
+        return queryText;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        String input = null;
+
+        try {
+            if (!this.validate(parameters)) {
```

Comment:
Move this `if` block out of `try` block, or change line 115 to `listener.onFailure(new ...); return;`

Reply:
if moving out of the try block, then when the input string is empty, it will throw illegalArgumentException directly.

Reply:
please see the expected test behavior here after the change for [empty input](https://github.com/opensearch-project/skills/pull/78/files#diff-b17f6e6e035a86ea534f4098c9b59d5df7c74180d6cff353c900d1dfc9e355e1R258), and [malformed input ](https://github.com/opensearch-project/skills/pull/78/files#diff-b17f6e6e035a86ea534f4098c9b59d5df7c74180d6cff353c900d1dfc9e355e1R265)